Clean steinberger material model

gassmoeller opened the pull request:
When looking over the Steinberger material model again, I realized its documentation is sub-par to our current standard (it was the first piece of code I contributed to ASPECT, 5 years ago). This PR cleans the code, and adds (backwards compatible) two options I need for a new model:
- It is now possible to provide one more material data file than there are compositional fields. For the often computed case of only one 'anomalous' material this allows to only compute with one field instead of two (thus saving 50% of computing time for compositional fields)
- Surprisingly to me the thermal conductivity was hard-coded. This is an input parameter now (same default as the previously hard-coded value).

All existing tests should continue to work, and I added a new test that uses the two options above.

tjhei commented on source/material_model/steinberger.cc:
```diff
@@ -501,21 +500,41 @@ namespace aspect
         {
-          if (n_material_data == 1)
-            cp = material_lookup[0]->dHdT(temperature,pressure);
+          if (material_file_names.size() == 1)
```
To me it looks like you won't need this special case but you can handle it in the ``else`` branch below (if you only do the sum if you have ``n_c+1`` files). This holds in all other places too.

gassmoeller replied:
Well, I could create the following:
```
if ((material_lookup.size() == 1) ||
    (material_lookup.size() == compositional_fields.size() + 1))
  {
            const double background_cp = material_lookup[0]->specific_heat(temperature,pressure);
            cp = background_cp;
            if (material_lookup.size() == compositional_fields.size() + 1)
              for (unsigned int i = 0; i < compositional_fields.size(); ++i)
                cp += compositional_fields[i] *
                      (material_lookup[i+1]->specific_heat(temperature,pressure) - background_cp);
    }
```
But that would only remove one duplicated line of assignment and trade it for an additional line of if-statement. Do you think this is easier to understand? I feel my version version is easier, because the simplest case (`material_lookup.size() == 1`) only has a single instruction.

I could also move the additional if-statement of the alternative above into an additional exit condition in the for-loop, but that would make the logic even more obscure.

tjhei commented on source/material_model/steinberger.cc:
```diff
@@ -501,21 +500,41 @@ namespace aspect
         {
-          if (n_material_data == 1)
-            cp = material_lookup[0]->specific_heat(temperature,pressure);
+          if (material_file_names.size() == 1)
```
is there a reason you are using the filenames instead of the size of ``material_lookup``?

tjhei commented on source/material_model/steinberger.cc:
```diff
@@ -715,10 +790,21 @@ namespace aspect
+                             "exactly as many files as compositional fields, the fields are "
+                             "assumed to represent the fractions of different materials "
+                             "and the average property is computed as a sum of "
+                             "the value of the compositional field times the"
```
missing space

tjhei commented on source/material_model/steinberger.cc:
```diff
@@ -789,11 +879,21 @@ namespace aspect
 
+        // Do some error checking
+        AssertThrow ((material_file_names.size() == this->n_compositional_fields()) ||
+            (material_file_names.size() == this->n_compositional_fields() + 1),
```
Didn't you also want to support a single file?

tjhei commented on data/material-model/steinberger/test-steinberger-compressible/testdata2.txt:
```diff
@@ -0,0 +1,10214 @@
+|6.6.6
```
maybe you have a more descriptive filename?

gassmoeller commented:
I addressed the comments, except for the one I commented on. I do not see how I could handle this with simpler logic (although in the beginning I also thought it could be handled in two cases, but I forgot the case where you have only one data file, but multiple compositional fields that you want to ignore).

tjhei commented:
> n. I do not see how I could handle this with simpler logic

would something like the following work?
```
int idx = 0;
double cp = (material_lookup.size() != compositional_fields.size()) ? 
    material_lookup[idx++]->dHdT(temperature,pressure)
       : 0.0;

for (unsigned int i = 0; i < material_lookup.size()-1; ++i)
  cp += compositional_fields[i] *
        (material_lookup[i+idx]->dHdT(temperature,pressure) - background_cp);
```

gassmoeller commented:
That version does not cover the case where `material_lookup.size() == compositional_fields.size() == 1`, in which case I want to get `cp = material_lookup[0]->dHdT(temperature,pressure)`. Additionally, I feel the logic is harder to understand than my explicit version. There might be a way to express the logic shorter, I just do not feel there is a way that is still simple to understand.

tjhei commented:
> Additionally, I feel the logic is harder to understand than my explicit version.

It was just a suggestion, so if you think it won't work or look more complicated, you are welcome to leave it as is.

gassmoeller commented:
> It was just a suggestion, so if you think it won't work or look more complicated, you are welcome to leave it as is.

Then I would prefer the current version :smile:. I updated the test result and addressed the other comments. From my point of view this is good to go.
